File: python-sdk/src/forta_agent/jwt.py
import os
import requests
import datetime
import json
import base64
import logging
import time
from web3.auto import w3
from web3 import Web3

MOCK_JWT = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJib3QtaWQiOiIweDEzazM4N2IzNzc2OWNlMjQyMzZjNDAzZTc2ZmMzMGYwMWZhNzc0MTc2ZTE0MTZjODYxeWZlNmMwN2RmZWY3MWYiLCJleHAiOjE2NjAxMTk0NDMsImlhdCI6MTY2MDExOTQxMywianRpIjoicWtkNWNmYWQtMTg4NC0xMWVkLWE1YzktMDI0MjBhNjM5MzA4IiwibmJmIjoxNjYwMTE5MzgzLCJzdWIiOiIweDU1NmY4QkU0MmY3NmMwMUY5NjBmMzJDQjE5MzZEMmUwZTBFYjNGNEQifQ.9v5OiiYhDoEbhZ-abbiSXa5y-nQXa104YCN_2mK7SP0'


def fetch_jwt(claims={}, expiresAt=None) -> str:
    if(os.environ['NODE_ENV'] != 'production'):
        return MOCK_JWT

    host_name = 'forta-jwt-provider'
    port = 8515
    path = '/create'
    uri = 'http://{host_name}:{port}{path}'.format(
        host_name=host_name, port=port, path=path)

    if((expiresAt != None) and (isinstance(expiresAt, datetime.datetime) == False)):
        raise Exception("expireAt must be of type datetime")

    if expiresAt is not None:
        exp_in_sec = expiresAt.timestamp()
        claims["exp"] = exp_in_sec

    response = requests.request("POST", uri, json={'claims': claims})

    if response.status_code == 200:
        data = response.json()
        return data.get('token')
    else:
        raise Exception(
            "Error occured with response fetching jwt token.", response)


def verify_jwt(token: str, polygonUrl: str = 'https://polygon-rpc.com') -> bool:
    splitJwt = token.split('.')
    rawHeader = splitJwt[0]
    rawPayload = splitJwt[1]

    header = json.loads(base64.urlsafe_b64decode(
        rawHeader + '==').decode('utf-8'))
    payload = json.loads(base64.urlsafe_b64decode(
        rawPayload + '==').decode('utf-8'))

    alg = header['alg']
    botId = payload['bot-id']
    expiresAt = payload['exp']
    signerAddress = payload['sub']

    if (signerAddress is None) or (botId is None):
        logging.warning('Invalid claim')
        return False

    if alg != 'ETH':
        logging.warning('Unexpected signing method: {alg}'.format(alg=alg))
        return False

    currentUnixTime = time.mktime(datetime.datetime.utcnow().utctimetuple())

    if expiresAt < currentUnixTime:
        logging.warning('Jwt expired')
        return False

    msg = '{header}.{payload}'.format(header=rawHeader, payload=rawPayload)
    msgHash = w3.keccak(text=msg)
    b64signature = splitJwt[2]
    signature = base64.urlsafe_b64decode(f'{b64signature}=').hex()
    recoveredSignerAddress = w3.eth.account.recoverHash(
        msgHash, signature=signature)

    if recoveredSignerAddress != signerAddress:
        logging.warn('Signature invalid: expected={signerAddress}, got={recoveredSignerAddress}'.format(
            signerAddress=signerAddress, recoveredSignerAddress=recoveredSignerAddress))
        return False

    w3Client = Web3(Web3.HTTPProvider(polygonUrl))
    DISPTACHER_ABI = [{"inputs": [{"internalType": "uint256", "name": "agentId", "type": "uint256"}, {"internalType": "uint256", "name": "scannerId", "type": "uint256"}],
                       "name": "areTheyLinked", "outputs": [{"internalType": "bool", "name": "", "type": "bool"}], "stateMutability": "view", "type": "function"}]
    # Source: https://docs.forta.network/en/latest/smart-contracts/
    DISPATCH_CONTRACT = "0xd46832F3f8EA8bDEFe5316696c0364F01b31a573"
    contract = w3Client.eth.contract(
        address=DISPATCH_CONTRACT, abi=DISPTACHER_ABI)

    areTheyLinked = contract.functions.areTheyLinked(
        int(botId, 0), int(recoveredSignerAddress, 0)).call()

    return areTheyLinked


class DecodedJwt:
    def __init__(self, dict):
        self.header = dict.get('header')
        self.payload = dict.get('payload')


def decode_jwt(token):
    # Adding need 4 byte for pythons b64decode
    header = base64.urlsafe_b64decode(
        token.split('.')[0] + '==').decode('utf-8')
    payload = base64.urlsafe_b64decode(
        token.split('.')[1] + '==').decode('utf-8')

    return DecodedJwt({
        "header": header,
        "payload": payload
    })

File: sdk/jwt.ts
import { ethers } from "ethers";
import { toUtf8Bytes } from "@ethersproject/strings";
import axios, { AxiosInstance } from "axios";
import { assertExists } from "./utils";

interface DecodedJwt {
  header: any;
  payload: any;
}

export const MOCK_JWT =
  "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJib3QtaWQiOiIweDEzazM4N2IzNzc2OWNlMjQyMzZjNDAzZTc2ZmMzMGYwMWZhNzc0MTc2ZTE0MTZjODYxeWZlNmMwN2RmZWY3MWYiLCJleHAiOjE2NjAxMTk0NDMsImlhdCI6MTY2MDExOTQxMywianRpIjoicWtkNWNmYWQtMTg4NC0xMWVkLWE1YzktMDI0MjBhNjM5MzA4IiwibmJmIjoxNjYwMTE5MzgzLCJzdWIiOiIweDU1NmY4QkU0MmY3NmMwMUY5NjBmMzJDQjE5MzZEMmUwZTBFYjNGNEQifQ.9v5OiiYhDoEbhZ-abbiSXa5y-nQXa104YCN_2mK7SP0";

export type FetchJwt = (claims: object, expiresAt?: Date) => Promise<string>;
export const provideFetchJwt = (axios: AxiosInstance): FetchJwt => {
  assertExists(axios, "axios");

  return async function fetchJwt(claims: object, expiresAt?: Date) {
    if (process.env.NODE_ENV !== "production") return MOCK_JWT;

    const hostname = "forta-jwt-provider";
    const port = 8515;
    const path = "/create";

    let fullClaims = { ...claims };

    if (expiresAt) {
      const expInSec = Math.floor(expiresAt.getTime() / 1000);

      // This covers the edge case where a Date that causes a seconds value to have number overflow resulting in a null exp
      const safeExpInSec =
        expInSec > Number.MAX_SAFE_INTEGER ? Number.MAX_SAFE_INTEGER : expInSec;

      fullClaims = {
        exp: safeExpInSec,
        ...fullClaims,
      };
    }

    const response = await axios.post(`http://${hostname}:${port}${path}`, {
      claims: fullClaims,
    });
    return response.data.token;
  };
};
export const fetchJwt = provideFetchJwt(axios);

export const decodeJwt = (token: string): DecodedJwt => {
  const splitJwt = token.split(".");
  const header = JSON.parse(Buffer.from(splitJwt[0], "base64").toString());
  const payload = JSON.parse(Buffer.from(splitJwt[1], "base64").toString());

  return {
    header,
    payload,
  };
};

export const verifyJwt = async (
  token: string,
  polygonRpcUrl: string = "https://polygon-rpc.com"
): Promise<boolean> => {
  const splitJwt = token.split(".");
  const rawHeader = splitJwt[0];
  const rawPayload = splitJwt[1];

  const header = JSON.parse(Buffer.from(rawHeader, "base64").toString());
  const payload = JSON.parse(Buffer.from(rawPayload, "base64").toString());

  const botId = payload["bot-id"] as string;
  const expiresAt = payload["exp"] as number;
  const algorithm = header?.alg;

  if (algorithm !== "ETH") {
    console.warn(`Unexpected signing method: ${algorithm}`);
    return false;
  }

  if (!botId) {
    console.warn(`Invalid claim`);
    return false;
  }

  const signerAddress = payload?.sub as string | undefined; // public key should be contract address that signed the JWT

  if (!signerAddress) {
    console.warn(`Invalid claim`);
    return false;
  }

  const currentUnixTime = Math.floor(Date.now() / 1000);

  if (expiresAt < currentUnixTime) {
    console.warn(`Jwt is expired`);
    return false;
  }

  const digest = ethers.utils.keccak256(
    toUtf8Bytes(`${rawHeader}.${rawPayload}`)
  );
  const signature = `0x${Buffer.from(splitJwt[2], "base64").toString("hex")}`;

  const recoveredSignerAddress = ethers.utils.recoverAddress(digest, signature); // Contract address that signed message

  if (recoveredSignerAddress !== signerAddress) {
    console.warn(
      `Signature invalid: expected=${signerAddress}, got=${recoveredSignerAddress}`
    );
    return false;
  }

  const polygonProvider = new ethers.providers.JsonRpcProvider(polygonRpcUrl);

  const DISPTACHER_ARE_THEY_LINKED =
    "function areTheyLinked(uint256 agentId, uint256 scannerId) external view returns(bool)";
  const DISPATCH_CONTRACT = "0xd46832F3f8EA8bDEFe5316696c0364F01b31a573"; // Source: https://docs.forta.network/en/latest/smart-contracts/
  const dispatchContract = new ethers.Contract(
    DISPATCH_CONTRACT,
    [DISPTACHER_ARE_THEY_LINKED],
    polygonProvider
  );
  const areTheyLinked = await dispatchContract.areTheyLinked(
    botId,
    recoveredSignerAddress
  );

  return areTheyLinked;
};
